### PR TITLE
Propagate trace inforation from incoming headers

### PR DESCRIPTION
**Pull request by evhen14**
# Description

The change allows Pub/Sub users to tap into tracing without requiring to publish CloudEvents on the subscription side.

The change assumes that `traceparent` and `tracestate` are the metadata fields used for this purpose. For backward compatibility, it also adds `traceid` field to the internally created cloud event.

## Issue reference

#7918 

## Checklist

Please make sure you've  completed the relevant tasks for this PR, out of the following list:

* [x] Code compiles correctly
* [x] Created/updated tests
* [x] Unit tests passing
* [ ] End-to-end tests passing
* [x] Extended the documentation / Created issue in the https://github.com/dapr/docs/ repo: dapr/docs#4278
* [ ] Specification has been updated / Created issue in the https://github.com/dapr/docs/ repo: dapr/docs#_[issue number]_
* [ ] Provided sample for the feature / Created issue in the https://github.com/dapr/docs/ repo: dapr/docs#_[issue number]_

### CODE REVIEW

File: docs/development/setup-dapr-development-env.md [hunk 1]
```diff
@@ -27,7 +27,7 @@ This document helps you get started developing Dapr. If you find any problems wh
 
 2. Install [Delve](https://github.com/go-delve/delve/tree/master/Documentation/installation) for Go debugging, if desired.
 
-3. Install [golangci-lint](https://golangci-lint.run/usage/install) version 1.55.2.
+3. Install [golangci-lint](https://golangci-lint.run/welcome/install/) version 1.55.2.
```
Comment (daixiang0) [inline docs/development/setup-dapr-development-env.md:30]: Seems it is not relative to the main topic, please update it in other PR, one thing one PR is good.

File: pkg/runtime/subscription/subscription.go [hunk 2]
```diff
@@ -162,7 +169,9 @@ func New(opts Options) (*Subscription, error) {
 				return err
 			}
 		} else {
+			// all messages consumed with "rawPayload=false" are deserialized as a CloudEvent, even when the payload is not a CloudEvent
```
Comment (daixiang0) [inline pkg/runtime/subscription/subscription.go:172]: Is it documented?

File: docs/development/setup-dapr-development-env.md (hunk 1 above)
Comment (yevgen-el8) [reply docs/development/setup-dapr-development-env.md:30]: I'll remove this change, but please don't expect a separate PR from me.

File: pkg/runtime/subscription/subscription.go (hunk 2 above)
Comment (yevgen-el8) [reply pkg/runtime/subscription/subscription.go:172]: What would be the best place to document it?

File: docs/development/setup-dapr-development-env.md (hunk 1 above)
Comment (yevgen-el8) [reply docs/development/setup-dapr-development-env.md]: Just to be clear, the old link is broken

Comment (evhen14) [reply docs/development/setup-dapr-development-env.md]: just to be clear, the old link is broken. I've removed this change, but please don't expect a separate PR from me.

File: pkg/runtime/subscription/subscription.go (hunk 2 above)
Comment (passuied) [reply pkg/runtime/subscription/subscription.go:172]: this is an implementation detail tbh. It doesn't mean that they will be sent as CloudEvent. it's just that they are being deserialized as cloud events regardless of whether they are rawPayload = true or false.

File: pkg/runtime/subscription/subscription.go [hunk 3]
```diff
@@ -175,6 +186,24 @@ func New(opts Options) (*Subscription, error) {
...
+
+			// fallback to message metadata to propagate the tracing information
+			if _, ok := cloudEvent[contribpubsub.TraceIDField]; !ok {
+				if traceid, ok := msg.Metadata[contribpubsub.TraceIDField]; ok {
+					cloudEvent[contribpubsub.TraceIDField] = traceid
+				}
+			}
+			if _, ok := cloudEvent[contribpubsub.TraceParentField]; !ok {
+				if traceparent, ok := msg.Metadata[contribpubsub.TraceParentField]; ok {
+					cloudEvent[contribpubsub.TraceIDField] = traceparent
```
Comment (yaron2) [inline pkg/runtime/subscription/subscription.go:198]: This seems to be different than the update logic above, where the TraceID populates both TraceID and TraceParent. In this block TraceParent is used for TraceID. Is this intentional?

Comment (evhen14) [reply pkg/runtime/subscription/subscription.go]: The `traceparent` override is based on [this comment](https://github.com/dapr/dapr/blob/master/pkg/runtime/pubsub/cloudevents.go#L51). I've made a change to make them consistent and adjusted the unit test

File: pkg/runtime/subscription/subscription.go [hunk 4]
```diff
@@ -175,6 +187,26 @@ func New(opts Options) (*Subscription, error) {
...
+			if _, ok := cloudEvent[contribpubsub.TraceIDField]; !ok {
+				if traceid, ok := msg.Metadata[contribpubsub.TraceIDField]; ok {
+					cloudEvent[contribpubsub.TraceIDField] = traceid
+				}
+			}
+			if _, ok := cloudEvent[contribpubsub.TraceParentField]; !ok {
+				if traceparent, ok := msg.Metadata[contribpubsub.TraceParentField]; ok {
+					cloudEvent[contribpubsub.TraceParentField] = traceparent
+					// traceparent supersedes traceid
+					cloudEvent[contribpubsub.TraceIDField] = traceparent
```
Comment (yaron2) [inline pkg/runtime/subscription/subscription.go:201]: This is repeating twice now

**Review by yaron2 — CHANGES_REQUESTED** [review]

**Review by yaron2 — APPROVED** [review]

**Pull request merged**